### DISCUSSION TITLE
Draw a roi in qskcontrol

### DISCUSSION THREAD

**Discussion by hestrro** [uwerat/qskinny#463] [Q&A] [answered]
When I draw a rectangle and follow the mouse to move the rectangle. I used two methods.
When I add two to QskWindow, both are displayed and moved correctly according to the correct coordinates of the paint function. However, when adding QskControl or QskItem, because QQuickPaintedItem does not have a setGeometry function, it cannot be displayed normally. Is there any solution?

1.继承一个QQuickPaintedItem
class GeometryItem:public QQuickPaintedItem
{
public:
    GeometryItem(QQuickItem *parent=nullptr);
    void paint(QPainter *painter) override;
    QRectF boundingRect() const override;
protected:
    void mousePressEvent(QMouseEvent *event) override;
    void mouseMoveEvent(QMouseEvent *event) override;
    void mouseReleaseEvent(QMouseEvent *event) override;
    double distance(const QPointF& point1, const QPointF& point2);
private:
    QPointF start;
    QPointF end;
    QPointF center;
    QRectF m_rect;
    bool ismoving=false;
    QPointF lastPostion;

};
2.继承一个QskGraphicLabel
class GeometryLaebl:public QskGraphicLabel
{
public:
    GeometryLaebl(QQuickItem *parent=nullptr);

    void updateGeometry();
protected:
    void mousePressEvent(QMouseEvent *event) override;
    void mouseMoveEvent(QMouseEvent *event) override;
    void mouseReleaseEvent(QMouseEvent *event) override;
    double distance(const QPointF& point1, const QPointF& point2);
    void paint();
private:
    QPainter painter;
    QRect boundrect;
    QPointF start;
    QPointF end;
    QPointF center;
    QRectF m_rect;
    bool ismoving=false;
    QPointF lastPostion;

};
 auto geometryPaintPage=new QskControl();
    geometryPaintPage->setMargins(0);
    addTab("Geometry_QQuickPaintedItem",geometryPaintPage);
    auto rect3=new GeometryItem(geometryPaintPage);

**Comment by hestrro** [comment]
When I set the size to 800, 800, the rectangle inside can display and drag the area of ​​800, 800 normally.
GeometryItem::GeometryItem(QQuickItem *parent):QQuickPaintedItem(parent)
{
    m_rect=QRect(0,0,200,200);
    setAcceptedMouseButtons(Qt::LeftButton);
    setSize(QSize(800,800));
    update();
    qDebug()<<textureSize();
}

**Comment by uwerat** [comment, marked as answer]
First of all: you don't need to create an image on the CPU and upload it as texture - what happens when using QQuickPaintedItem - for displaying a rectangle. The most simple solution would be to use a QskControl and use QskControl::setBackgroundColor or QskControl::setBackground instead.

QQuickItem has methods like setX, setY, setSize,  where you can change its geometry. If you want to do it in one call: in QskQuick.h, you will find qskSetItemGeometry.

In general I recommend to read some documentation about how the Qt/Quick graphic stack works - f.e https://doc.qt.io/qt-6/qtquick-visualcanvas-scenegraph.html. Then it should be obvious why objects that live in the scene graph thread won't be updated immediately. Note that QQuickItem::update schedules a call to updatePaintNode without updating anything.

**Comment by uwerat** [comment]
By the way: please mark discussions as answered once you consider them so

**Comment by hestrro** [comment]
![rect](https://github.com/user-attachments/assets/cc6827fd-59c5-4549-afbc-c12d1eae9b7d)



When the mouse moves, the m_rect variable is updated, and then the paint function is called through update. QQuickPaintedItem automatically handles coordinate system conversion when it's inside a QQuickWindow, ensuring that what you draw in the paint function maps correctly to the item's local coordinate system. You don't need to manually handle coordinate system transformations. So when I put it in qskcontrol, I updated it manually through qskSetItemGeometry(this, m_rect);

but ROI will disappear after following the mouse beyond a certain area.




void GeometryItem::paint(QPainter *painter)
{
    painter->setRenderHint(QPainter::Antialiasing);

    QPen pen(Qt::green);
    pen.setWidth(1);
    painter->setPen(pen);
    painter->setBrush(Qt::NoBrush);

    QPainterPath path;
    path.addRect(m_rect);
    path.addEllipse(m_rect.center(), 10, 10);
    qskSetItemGeometry(this, m_rect);
    painter->drawPath(path);
}

**Reply by uwerat** [reply]
Sorry but this task is unrelated to QSkinny - it is something that fits on a Qt support channel.

**Comment by hestrro** [comment]
![image](https://github.com/user-attachments/assets/b5d262e1-ae48-4465-a0af-dc0ae1fddd3b)
However, it is indeed very convenient to realize the ROI of various graphics through qskgraphiclabel, and can well implement operations such as zooming in, zooming out, and panning. The same situation as QQuickPaintedItem will not occur through setGeomtry. And it's drawn to the gpu.
